File: 417-pacific-atlantic-water-flow/417-pacific-atlantic-water-flow.py
class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        ROWS, COLS = len(heights), len(heights[0])
        directions = [[1, 0], [-1, 0], [0, 1], [0, -1]]
        atlantic, pacific = set(), set()
        
        def bordersPacific(r, c):
            return r == 0 or c == 0
        
        def bordersAtlantic(r, c):
            return r == ROWS - 1 or c == COLS - 1
        
        def inOcean(r, c):
            return r in [-1, ROWS] or c in [-1, COLS]
        
        def dfs(r, c, prev, res):
            if inOcean(r, c) or (r, c) in res or heights[r][c] < prev:
                return 
            res.add((r, c))
            for dr, dc in directions:
                dfs(r + dr, c + dc, heights[r][c], res)
                
        for r in range(ROWS):
            for c in range(COLS):
                if bordersPacific(r, c):
                    dfs(r, c, -1, pacific)
                    
                if bordersAtlantic(r, c):
                    dfs(r, c, -1, atlantic)
                    
        res = []
        for r in range(ROWS):
            for c in range(COLS):
                if (r, c) in atlantic and (r, c) in pacific:
                    res.append([r, c])
                    
        return res
            

File: 417-pacific-atlantic-water-flow/417-pacific-atlantic-water-flow.java
class Solution {
    private static final int[][] DIRECTIONS = new int[][] {
        new int[] { 0, 1 },
        new int[] { 0, -1 },
        new int[] { 1, 0 },
        new int[] { -1, 0 }
    };
    
    private int[][] grid;
    
    public List<List<Integer>> pacificAtlantic(int[][] heights) {
        List<List<Integer>> res = new ArrayList<>();
        if (heights.length == 0) return res;
        
        grid = heights;
        Set<List<Integer>> flowsToAtlantic = new HashSet<>();
        Set<List<Integer>> flowsToPacific = new HashSet<>();
        for (int r = 0; r < grid.length; r++) {
            for (int c = 0; c < grid[0].length; c++) {
                if (bordersAtlantic(r, c)) 
                    dfs(r, c, 0, flowsToAtlantic);
                
                if (bordersPacific(r, c)) 
                    dfs(r, c, 0, flowsToPacific);
            }
        }
        
        for (List<Integer> coord : flowsToAtlantic) {
            if (flowsToPacific.contains(coord)) {
                res.add(coord);
            }
        }
        return res;
    }
    
    private void dfs(int r, int c, int prev, Set<List<Integer>> flowsToOcean) {
        if (isInOcean(r, c) || flowsToOcean.contains(Arrays.asList(r, c)) || grid[r][c] < prev) {
            return;
        }
        flowsToOcean.add(Arrays.asList(r, c));
        
        for (int[] direction : DIRECTIONS) {
            int dr = direction[0], dc = direction[1];
            dfs(r + dr, c + dc, grid[r][c], flowsToOcean);
        } 
    }
    
    private boolean isInOcean(int r, int c) {
        return r < 0 || r >= grid.length || c < 0 || c >= grid[0].length;
    }
    
    private boolean bordersPacific(int r, int c) {
        return r == 0 || c == 0;
    }
    
    private boolean bordersAtlantic(int r, int c) {
        return r == grid.length - 1 || c == grid[0].length - 1;
    }
}